### PR TITLE
Apply transposition when propagating through an elementwise operation

### PR DESCRIPTION
**Pull request by chelini**
Unpack has `transpose` semantics with `outer_dims_perm` when propagating through an operation we need to apply transposition if `outer_dims_perm` is available.

### CODE REVIEW

File: lib/TPP/ToBlockLayoutAndBack.cpp [hunk 1]
```diff
@@ -587,17 +601,38 @@ struct PropagateThroughElementWiseOp
     : public OpRewritePattern<linalg::GenericOp> {
   using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
 
-  // Check sinking preconditions: a) all loops must
-  // be parallel.
-  LogicalResult checkPreconditions(linalg::GenericOp linalgOp) const {
-    if (linalgOp.getNumLoops() != linalgOp.getNumParallelLoops())
+  LogicalResult isElementwise(linalg::GenericOp linalgOp) const {
+    if (!linalg::isElementwise(linalgOp))
```
Comment (rengolin) [inline lib/TPP/ToBlockLayoutAndBack.cpp:605]: This looks like a nop function, why not just use `linalg::isElementwise` directly?

File: lib/TPP/ToBlockLayoutAndBack.cpp [hunk 2]
```diff
@@ -587,17 +601,38 @@ struct PropagateThroughElementWiseOp
...
+                        [](AffineMap map) { return map.isMinorIdentity(); });
+  }
+
+  // Require operands to come from a single `unpack` operation.
+  LogicalResult hasOnlyOnePackedOperand(linalg::GenericOp linalgOp) const {
+    bool oneOf = false;
+    for (OpOperand &operand : linalgOp->getOpOperands()) {
+      linalgx::UnPackOp unpackOp =
+          operand.get().getDefiningOp<linalgx::UnPackOp>();
+      if (unpackOp) {
```
Comment (rengolin) [inline lib/TPP/ToBlockLayoutAndBack.cpp:622]: I'm guessing there aren't many operands, and this logic while meaningful, is a bit convoluted.

Probably be simpler to just count and `return count != 1`?

File: lib/TPP/ToBlockLayoutAndBack.cpp [hunk 3]
```diff
@@ -703,6 +789,19 @@ struct PropagateThroughElementWiseOp
         unpackOutputs.push_back(unpackOp.getOutput());
     }
 
+    // XXX: innerDimsPos and outerDimsPerm must have been unsigned in the first
```
Comment (rengolin) [inline lib/TPP/ToBlockLayoutAndBack.cpp:792]: What does XXX mean?

Comment (chelini) [reply lib/TPP/ToBlockLayoutAndBack.cpp:792]: Something bad that I would like to fix later. But essentially the problem is that we should have designed `pack` and `unpack` with unsigned for `inner_dims` and `outer_dims_perms`.

File: lib/TPP/ToBlockLayoutAndBack.cpp (hunk 2 above)
Comment (chelini) [reply lib/TPP/ToBlockLayoutAndBack.cpp]: Done.

File: lib/TPP/ToBlockLayoutAndBack.cpp (hunk 1 above)
Comment (chelini) [reply lib/TPP/ToBlockLayoutAndBack.cpp]: Done.

File: lib/TPP/ToBlockLayoutAndBack.cpp (hunk 3 above)
Comment (rengolin) [reply lib/TPP/ToBlockLayoutAndBack.cpp:782]: Then FIXME? or TODO?

Comment (chelini) [reply lib/TPP/ToBlockLayoutAndBack.cpp]: Done.

**Comment by chelini** [comment]
Any more comments here?

**Review by rengolin — APPROVED** [review]

**Pull request merged**